mlir/lib/Quantum/Utils/QuantumSplitting.cpp

```diff
@@ -89,17 +91,95 @@ void AugmentedCircuitGenerator::generate(Region &region, OpBuilder &builder)
         else if (auto extractOp = dyn_cast<quantum::ExtractOp>(op)) {
             cacheDynamicWire(extractOp, builder);
         }
-        else if (auto gate = dyn_cast<quantum::DifferentiableGate>(op)) {
-            ValueRange diffParams = gate.getDiffParams();
-            if (!diffParams.empty()) {
-                for (Value param : diffParams) {
-                    builder.create<ListPushOp>(gate.getLoc(), oldToCloned.lookupOrDefault(param),
-                                               cache.paramVector);
+        else if (auto gate = dyn_cast<quantum::ParametrizedGate>(op)) {
+            ValueRange params = gate.getAllParams();
+
+            for (Value param : params) {
+                Location loc = gate.getLoc();
+                Value clonedParam = oldToCloned.lookupOrDefault(param);
+                Type paramType = clonedParam.getType();
+
+                if (paramType.isF64()) {
+                    builder.create<ListPushOp>(loc, clonedParam, cache.paramVector);
+                    continue;
+                }
+
+                // Sanitizing inputs.
+                // Technically we know for a fact that none of this will ever issue an error.
+                // This is because QubitUnitary is guaranteed to have a tensor<NxNxcomplex<f64>>
+                // But this code in the future may be extended to support other types.
+                // Hence the sanitization.
+                if (!isa<RankedTensorType>(paramType)) {
+                    gate.emitOpError() << "Unexpected type.";
+                }
+
+                auto aTensor = paramType.cast<RankedTensorType>();
+                ArrayRef<int64_t> shape = aTensor.getShape();
+
+                if (2 != shape.size()) {
+                    gate.emitOpError() << "Unexpected tensor shape in QubitUnitaryOp";
```

Comment:
The references to QubitUnitaryOps are a bit misplaced here, since we are working based on the parametrized gate interface and the types of the parameters.
```suggestion
                    gate.emitOpError() << "Caching only supports rank 2 tensors";
```

Reply:
Ok, I understand, you want this to be generalized. My perception of this is that we really cared about building the wheels sometime today and that we agreed that we would work on this later on to improve the interface. Hence why I didn't put too much importance in generalizing it. I will generalize it.